pkg/kubelet/volumemanager/cache/actual_state_of_world.go

```diff
@@ -346,6 +350,110 @@ func (asw *actualStateOfWorld) MarkVolumeAsDetached(
 	asw.DeleteVolume(volumeName)
 }
 
+func (asw *actualStateOfWorld) IsVolumeReconstructed(volumeName v1.UniqueVolumeName, podName volumetypes.UniquePodName) bool {
+	volumeState := asw.GetVolumeMountState(volumeName, podName)
+
+	// only uncertain volumes are reconstructed
+	if volumeState != operationexecutor.VolumeMountUncertain {
+		return false
+	}
+
+	asw.RLock()
+	defer asw.RUnlock()
+	podMap, ok := asw.foundDuringReconstruction[volumeName]
+	if !ok {
+		return false
+	}
+	_, foundPod := podMap[podName]
+	return foundPod
+}
+
+func (asw *actualStateOfWorld) CheckAndMarkVolumeAsUncertainViaReconstruction(opts operationexecutor.MarkVolumeOpts) (bool, error) {
+	asw.Lock()
+	defer asw.Unlock()
+
+	volumeObj, volumeExists := asw.attachedVolumes[opts.VolumeName]
+	if !volumeExists {
+		return false, nil
+	}
+
+	podObj, podExists := volumeObj.mountedPods[opts.PodName]
+	if podExists {
+		// if volume mount was uncertain we should keep trying to unmount the volume
+		if podObj.volumeMountStateForPod == operationexecutor.VolumeMountUncertain {
+			return false, nil
+		}
+		if podObj.volumeMountStateForPod == operationexecutor.VolumeMounted {
+			return false, nil
+		}
+	}
+
+	podName := opts.PodName
+	podUID := opts.PodUID
+	volumeName := opts.VolumeName
+	mounter := opts.Mounter
+	blockVolumeMapper := opts.BlockVolumeMapper
+	outerVolumeSpecName := opts.OuterVolumeSpecName
+	volumeGidValue := opts.VolumeGidVolume
+	volumeSpec := opts.VolumeSpec
+
+	if !podExists {
+		podObj = mountedPod{
+			podName:                podName,
+			podUID:                 podUID,
+			mounter:                mounter,
+			blockVolumeMapper:      blockVolumeMapper,
+			outerVolumeSpecName:    outerVolumeSpecName,
+			volumeGidValue:         volumeGidValue,
+			volumeSpec:             volumeSpec,
+			volumeMountStateForPod: opts.VolumeMountState,
+		}
+	}
+
+	// If pod exists, reset remountRequired value
+	podObj.remountRequired = false
+	podObj.volumeMountStateForPod = opts.VolumeMountState
+
+	// if volume is mounted successfully, then it should be removed from foundDuringReconstruction map
+	if opts.VolumeMountState == operationexecutor.VolumeMounted {
```

Comment:
for now, I see opts.VolumeMountState should always be uncertain, right?

Reply:
yep. hardcoded the state in this function to be `VolumeMountUncertain`.